modules/market/src/main/java/com/opengamma/strata/market/key/QuotesArrayKey.java

```diff
@@ -0,0 +1,420 @@
+ */
+package com.opengamma.strata.market.key;
+
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import org.joda.beans.Bean;
+import org.joda.beans.BeanDefinition;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaProperty;
+import org.joda.beans.Property;
+import org.joda.beans.PropertyDefinition;
+import org.joda.beans.impl.direct.DirectFieldsBeanBuilder;
+import org.joda.beans.impl.direct.DirectMetaBean;
+import org.joda.beans.impl.direct.DirectMetaProperty;
+import org.joda.beans.impl.direct.DirectMetaPropertyMap;
+
+import com.opengamma.strata.basics.market.FieldName;
+import com.opengamma.strata.basics.market.MarketDataBox;
+import com.opengamma.strata.basics.market.ScenarioMarketDataKey;
+import com.opengamma.strata.collect.array.DoubleArray;
+import com.opengamma.strata.collect.id.StandardId;
+import com.opengamma.strata.market.value.QuotesArray;
+
+/**
+ * A key identifying a {@link QuotesArray} containing values for a piece of quoted market data in multiple scenarios.
+ */
+@BeanDefinition
+public final class QuotesArrayKey
+    implements ScenarioMarketDataKey<Double, QuotesArray>, ImmutableBean {
+
+  /**
+   * The ID of the market data that is required, typically an ID from an external data provider.
+   * For example, 'Bloomberg~AAPL'.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final StandardId standardId;
+
+  /**
+   * The field name in the market data record that is required.
+   * For example, {@link FieldName#MARKET_VALUE}.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final FieldName fieldName;
+
+  /**
+   * Returns a key identifying the market data with the specified ID and field name.
+   *
+   * @return a key identifying the market data with the specified ID and field name
+   */
+  public static QuotesArrayKey of(StandardId id, FieldName fieldName) {
+    return new QuotesArrayKey(id, fieldName);
+  }
+
+  /**
+   * Returns a key identifying the same market data as the quote key.
+   *
+   * @return a key identifying the same market data as the quote key
+   */
+  public static QuotesArrayKey of(QuoteKey quoteKey) {
+    return new QuotesArrayKey(quoteKey.getStandardId(), quoteKey.getFieldName());
+  }
+
+  @Override
+  public QuoteKey getMarketDataKey() {
+    return QuoteKey.of(standardId, fieldName);
+  }
+
+  @Override
+  public Class<QuotesArray> getScenarioMarketDataType() {
+    return QuotesArray.class;
+  }
+
+  @Override
+  public QuotesArray createScenarioValue(MarketDataBox<Double> marketDataBox) {
+    // TODO This suggests MarketDataBox needs a stream() method or at least a helper to create a stream from a box
+    double[] quotes = new double[marketDataBox.getScenarioCount()];
+
+    for (int i = 0; i < marketDataBox.getScenarioCount(); i++) {
+      quotes[i] = marketDataBox.getValue(i);
+    }
+    return QuotesArray.of(DoubleArray.ofUnsafe(quotes));
```

Comment:
return QuotesArray.of(DoubleArray.of(marketDataBox.getScenarioCount(), i -> marketDataBox.getValue(i)));